cli: add --system param to allow validating system user-data on a machine

blackboxsw opened the pull request:
## Proposed Commit Message
> cli: add --system param schema to allow validating system user-data
>
> Allow root user to validate the userdata provided to the launched a machine using:
> cloud-init devel schema --system

## Test Steps
```bash
lxc launch ubuntu-daily:xenial dev-x
lxc file push cloudinit/config/schema.py dev-x/usr/lib/python3/dist-packages/cloudinit/config/
lxc exec dev-x bash
$ cloud-init devel schema --system
```

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes
that apply. -->
 - [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/topics/hacking.html)
 - [x] I have updated or added any unit tests accordingly
 - [x] I have updated or added any documentation accordingly: **NOTE:** separating auto-generated documentation for all CLI commands to a separate PR

mitechie commented on cloudinit/config/schema.py:
```diff
@@ -425,6 +443,8 @@ def get_parser(parser=None):
     parser.add_argument('-c', '--config-file',
                         help='Path of the cloud-config yaml file to validate')
+    parser.add_argument('--system', action='store_true', default=False,
+                        help='Validate the system cloud-config userdata')
```
shouldn't there be associated doc updates with adding a new argument to UI?

mitechie commented on doc/rtd/topics/faq.rst:
```diff
@@ -135,12 +135,12 @@ that can validate your user data offline.
-Another option is to run the following on an instance when debugging:
+Another option is to run the following on an instance to debug userdata
+provided to the system:
 
```
This is the doc update but it seemed an edit to an existing doc and didn't seem clear to me.

blackboxsw replied:
Yeah, devel commands, and general CLI documentation, isn't great. In the devel case I had originally left it fairly undocumented as these `devel` commands will likely change form or break compatibility before we decide to make them fully fledged top-level (non-devel) subcommands of cloud-init.  But, now that we are referencing subcommand switches in docs in other places. it would be helpful to have a more full rendering of cli docs for this devel subcommands. I ultimately think we should autogeerate docs from the argparse configs so that we can keep that up to date any time the cli is changed.  I think `sphinx-argparse`  extension could do this for us, but I'd like to leave that to a separate PR as it'll potentially involve re-organizing the top-level [CLI docs](https://cloudinit.readthedocs.io/en/latest/topics/cli.html) docs. As I think we could try to modify how the CLI docs in general are exposed and documented.

blackboxsw replied:
That said, I did just push a commit that renders cmdline devel schema to this branch. I think we can build on this if an acceptable format for all CLI commands.

OddBloke reviewed (changes requested):
Thanks Chad, this is a nice little UX improvement. A couple of inline comments.

blackboxsw commented on cloudinit/config/schema.py:
```diff
@@ -173,17 +176,32 @@ def annotated_cloudconfig_file(cloudconfig, original_content, schema_errors):
-        raise RuntimeError('Configfile {0} does not exist'.format(config_path))
-    content = load_file(config_path, decode=False)
+    if config_path == SYSTEM_USERDATA_FILE:
+        (content, err) = subp(
+            ["cloud-init", "query", "userdata"], decode=False
+        )
```
@OddBloke The reason we treat this specially is because the typical userdata path in some cases is actually a compressed binary (juju deployed units). In this case the query command uses the special [cloudinit.cmd.query.load_userdata func to decompress and decode](https://github.com/canonical/cloud-init/blob/master/cloudinit/cmd/query.py#L81-L93) that I wanted to leverage here. It would probably be preferable to either call the function directly or relocate that utility function  into util.py.

blackboxsw commented:
@OddBloke @mitechie I've updated this branch and added auto-generated docs for review for only the devel subcommand using sphinx-argparse. If this documentation approach is palatable. I can add a followup PR to auto-gen docs for our entire CLI as a followup PR. (I could separate that from this PR too).

blackboxsw commented:
@OddBloke @mitechie separated doc changes into a followup PR as discussed in standup yesterday

mitechie reviewed (approved):
With a +1 to Dan's comment on the file path at the start of the PR.

blackboxsw commented on cloudinit/config/schema.py:
```diff
@@ -39,6 +40,8 @@
 SCHEMA_EXAMPLES_HEADER = '\n**Examples**::\n\n'
 SCHEMA_EXAMPLES_SPACER_TEMPLATE = '\n    # --- Example{0} ---'
 
+SYSTEM_USERDATA_FILE = "/var/lib/cloud/instance/user-data.txt"
```
@mitechie @OddBloke OOPS while the code no longer relies on this static var, I had forgotten to remove it.
Done.

OddBloke commented on cloudinit/config/schema.py:
```diff
@@ -449,7 +468,11 @@ def handle_schema_args(name, args):
+                cfg_name = "system userdata"
+            else:
+                cfg_name = args.config_file
+            print("Valid cloud-config: {0}".format(cfg_name))
```
Obviously a nit, but `print` takes `*args` and prints them:

```suggestion
            print("Valid cloud-config:", cfg_name)
```

```python
In [1]: print("a", "b")                                                                                                                                                                                                                                                                   
a b
```

OddBloke commented on cloudinit/config/schema.py:
```diff
@@ -435,11 +454,11 @@ def get_parser(parser=None):
     """Handle provided schema args and perform the appropriate actions."""
-    exclusive_args = [args.config_file, args.docs]
+    exclusive_args = [args.config_file, args.docs, args.system]
     if not any(exclusive_args) or all(exclusive_args):
```
I think this check needs updating: `all(exclusive_args)` will be `False` if any two of the three are passed:

```py
In [3]: exclusive_args = ["config_file", None, "system"]                                                                                                                                                                                                                                  

In [4]: if not any(exclusive_args) or all(exclusive_args): 
   ...:     print('Expected one of --config-file, --system or --docs arguments') 
   ...:      
In [5]:
```

blackboxsw replied:
Good find. Changed logic and added unit tests

OddBloke reviewed (changes requested):
Thanks Chad! Overall this looks good, I have one inline comment that I think needs addressing before this can land (and another one which you can ignore because it's a nit :+1:).

OddBloke commented on tests/unittests/test_handler/test_schema.py:
```diff
@@ -400,50 +400,97 @@ def test_annotated_cloudconfig_file_annotates_separate_line_items(self):
+        )
+        assert expected == err
+
+    def test_main_missing_args(self, capsys):
```
A nit (unless I'm wrong, then not even that): you could potentially fold this into the above definition by making the parameters `list(exclusive_combinations) + [()]`.

OddBloke commented on tests/unittests/test_handler/test_schema.py:
```diff
@@ -400,50 +400,97 @@ def test_annotated_cloudconfig_file_annotates_separate_line_items(self):
+        ["--system", "--docs all", "--config-file something"], 2
+    )
+
+    @pytest.mark.parametrize("params", exclusive_combinations)
```
Nit: a not-particularly-important gap here is all three being specified at once.

OddBloke commented on cloudinit/config/schema.py:
```diff
@@ -173,17 +174,33 @@ def annotated_cloudconfig_file(cloudconfig, original_content, schema_errors):
-    content = load_file(config_path, decode=False)
+    if config_path is None:
+        # Use system's raw userdata path
+        if os.getuid() != 0:
```
This file is always root:root-owned by default and not world-readable, so this is not even a nit: we could instead rely on the user's permissions to determine if they can use this path.

OddBloke reviewed (approved):
Thanks Chad! Some nits inline, which you can fix in a follow-up if you so desire.
